File: sic-api/Controllers/Storage/StorageController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using sic_api.Controllers;
using sic_api.Features.Storage;
using sic_api.Model.Storage;
using sic_api.Services.Interfaces;

namespace sic_api.Controllers.Storage;

[Route("api/storage")]
public class StorageController(
    IResumableUploadService resumableUploadService,
    IFileStorageService fileStorageService) : BaseController
{
    [Authorize]
    [HttpPost("upload/image")]
    [RequestSizeLimit(200_000_000)]
    public async Task<IActionResult> UploadImage([FromForm] StorageUploadRequest request, CancellationToken cancellationToken)
    {
        var result = await Mediator.Send(new UploadStorageFile.Command
        {
            File = request.File,
            Visibility = request.Visibility,
            UploadGroupId = request.UploadGroupId,
            Category = FileCategory.Image,
            HttpContext = HttpContext
        }, cancellationToken);
        return Ok(new { id = result.Id, accessUrl = result.AccessUrl, fileName = result.FileName, fileSize = result.FileSize, contentType = result.ContentType });
    }

    [Authorize]
    [HttpPost("upload/video")]
    [RequestSizeLimit(1_000_000_000)]
    public async Task<IActionResult> UploadVideo([FromForm] StorageUploadRequest request, CancellationToken cancellationToken)
    {
        var result = await Mediator.Send(new UploadStorageFile.Command
        {
            File = request.File,
            Visibility = request.Visibility,
            UploadGroupId = request.UploadGroupId,
            Category = FileCategory.Video,
            HttpContext = HttpContext
        }, cancellationToken);
        return Ok(new { id = result.Id, accessUrl = result.AccessUrl, fileName = result.FileName, fileSize = result.FileSize, contentType = result.ContentType });
    }

    [Authorize]
    [HttpPost("upload/document")]
    [RequestSizeLimit(200_000_000)]
    public async Task<IActionResult> UploadDocument([FromForm] StorageUploadRequest request, CancellationToken cancellationToken)
    {
        var result = await Mediator.Send(new UploadStorageFile.Command
        {
            File = request.File,
            Visibility = request.Visibility,
            UploadGroupId = request.UploadGroupId,
            Category = FileCategory.Document,
            HttpContext = HttpContext
        }, cancellationToken);
        return Ok(new { id = result.Id, accessUrl = result.AccessUrl, fileName = result.FileName, fileSize = result.FileSize, contentType = result.ContentType });
    }

    [Authorize]
    [HttpPost("upload/sessions")]
    public async Task<IActionResult> CreateUploadSession(
        [FromBody] StorageUploadSessionCreateRequest request,
        CancellationToken cancellationToken) =>
        Ok(await resumableUploadService.CreateSessionAsync(request, cancellationToken));

    [Authorize]
    [HttpGet("upload/sessions/{sessionId:guid}")]
    public async Task<IActionResult> GetUploadSessionStatus(
        Guid sessionId,
        CancellationToken cancellationToken) =>
        Ok(await resumableUploadService.GetSessionStatusAsync(sessionId, cancellationToken));

    [Authorize]
    [HttpPost("upload/sessions/{sessionId:guid}/chunks/{chunkIndex:int}")]
    [RequestSizeLimit(100_000_000)]
    public async Task<IActionResult> UploadChunk(
        Guid sessionId,
        int chunkIndex,
        [FromForm] StorageUploadSessionChunkRequest request,
        CancellationToken cancellationToken) =>
        Ok(await resumableUploadService.UploadChunkAsync(sessionId, chunkIndex, request.Chunk, cancellationToken));

    [Authorize]
    [HttpPost("upload/sessions/{sessionId:guid}/complete")]
    public async Task<IActionResult> CompleteUploadSession(
        Guid sessionId,
        CancellationToken cancellationToken) =>
        Ok(await resumableUploadService.CompleteSessionAsync(sessionId, HttpContext, cancellationToken));

    [Authorize]
    [HttpDelete("upload/sessions/{sessionId:guid}")]
    public async Task<IActionResult> CancelUploadSession(
        Guid sessionId,
        CancellationToken cancellationToken)
    {
        await resumableUploadService.CancelSessionAsync(sessionId, cancellationToken);
        return NoContent();
    }

    [Authorize]
    [HttpPost("uploads/{uploadId:guid}/activate")]
    public async Task<IActionResult> ActivateUpload(
        Guid uploadId,
        CancellationToken cancellationToken)
    {
        await fileStorageService.ActivateAsync(uploadId, User, cancellationToken);
        return NoContent();
    }

    [AllowAnonymous]
    [HttpGet("files/{bucketName}/{**objectKey}")]
    public async Task<IActionResult> DownloadFile(
        string bucketName,
        string objectKey,
        [FromQuery] int? width,
        [FromQuery] int? height,
        CancellationToken cancellationToken)
        => await Mediator.Send(new DownloadStorageFile.Query
        {
            BucketName = bucketName,
            ObjectKey = objectKey,
            Width = width,
            Height = height,
            User = User
        }, cancellationToken);
}

File: sic-spring/sic/src/main/java/com/softinter/sicapi/controller/storage/StorageController.java
package com.softinter.sicapi.controller.storage;

import com.softinter.sicapi.dto.request.CompleteUploadSessionRequest;
import com.softinter.sicapi.dto.request.UploadSessionRequest;
import com.softinter.sicapi.dto.response.StorageDownloadResponse;
import com.softinter.sicapi.dto.response.StorageUploadResponse;
import com.softinter.sicapi.dto.response.UploadSessionResponse;
import com.softinter.sicapi.service.FileStorageService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.core.io.InputStreamResource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api/storage")
@RequiredArgsConstructor
@SecurityRequirement(name = "Bearer Authentication")
@Tag(name = "Storage", description = "File Storage API")
public class StorageController {

    private final FileStorageService fileStorageService;

    // ========== Single file upload ==========
    @PostMapping("/upload/image")
    public ResponseEntity<?> uploadImage(@RequestParam("file") MultipartFile file,
                                         @RequestParam(required = false) String visibility,
                                         @RequestParam(required = false) UUID uploadGroupId) {
        StorageUploadResponse res = fileStorageService.uploadFile(file, "IMAGE", visibility, uploadGroupId);
        return ResponseEntity.ok(Map.of(
                "id", res.getId(),
                "accessUrl", res.getFileUrl(),
                "fileName", res.getFileName(),
                "fileSize", res.getFileSize(),
                "contentType", res.getMimeType()
        ));
    }

    @PostMapping("/upload/video")
    public ResponseEntity<?> uploadVideo(@RequestParam("file") MultipartFile file,
                                         @RequestParam(required = false) String visibility,
                                         @RequestParam(required = false) UUID uploadGroupId) {
        StorageUploadResponse res = fileStorageService.uploadFile(file, "VIDEO", visibility, uploadGroupId);
        return ResponseEntity.ok(Map.of(
                "id", res.getId(),
                "accessUrl", res.getFileUrl(),
                "fileName", res.getFileName(),
                "fileSize", res.getFileSize(),
                "contentType", res.getMimeType()
        ));
    }

    @PostMapping("/upload/document")
    public ResponseEntity<?> uploadDocument(@RequestParam("file") MultipartFile file,
                                            @RequestParam(required = false) String visibility,
                                            @RequestParam(required = false) UUID uploadGroupId) {
        StorageUploadResponse res = fileStorageService.uploadFile(file, "DOCUMENT", visibility, uploadGroupId);
        return ResponseEntity.ok(Map.of(
                "id", res.getId(),
                "accessUrl", res.getFileUrl(),
                "fileName", res.getFileName(),
                "fileSize", res.getFileSize(),
                "contentType", res.getMimeType()
        ));
    }

    // ========== Resumable upload session ==========
    @PostMapping("/upload/sessions")
    @Operation(summary = "Create upload session")
    public ResponseEntity<UploadSessionResponse> createUploadSession(@RequestBody UploadSessionRequest request) {
        UploadSessionResponse response = fileStorageService.createUploadSession(request);
        return ResponseEntity.ok(response);
    }

    @GetMapping("/upload/sessions/{sessionId}")
    @Operation(summary = "Get session status")
    public ResponseEntity<Map<String, Object>> getSessionStatus(@PathVariable UUID sessionId) {
        // TODO: implement status response
        return ResponseEntity.ok(Map.of("sessionId", sessionId));
    }

    @PostMapping("/upload/sessions/{sessionId}/chunks/{chunkIndex}")
    @Operation(summary = "Upload a chunk")
    public ResponseEntity<UploadSessionResponse> uploadChunk(
            @PathVariable UUID sessionId,
            @PathVariable int chunkIndex,
            @RequestParam("chunk") MultipartFile chunk) {
        UploadSessionResponse response = fileStorageService.uploadChunk(sessionId, chunkIndex, chunk);
        return ResponseEntity.ok(response);
    }

    @PostMapping("/upload/sessions/{sessionId}/complete")
    @Operation(summary = "Complete upload session (using path variable)")
    public ResponseEntity<StorageUploadResponse> completeUploadSession(@PathVariable UUID sessionId,
                                                                       HttpServletRequest request) {
        StorageUploadResponse response = fileStorageService.completeUploadSession(sessionId, request);
        return ResponseEntity.ok(response);
    }

    @DeleteMapping("/upload/sessions/{sessionId}")
    @Operation(summary = "Cancel upload session")
    public ResponseEntity<Void> cancelSession(@PathVariable UUID sessionId) {
        fileStorageService.cancelSession(sessionId);
        return ResponseEntity.noContent().build();
    }

    @PostMapping("/uploads/{uploadId}/activate")
    @Operation(summary = "Activate upload")
    public ResponseEntity<Void> activateUpload(@PathVariable UUID uploadId) {
        fileStorageService.activateUpload(uploadId);
        return ResponseEntity.noContent().build();
    }

    // ========== Legacy endpoint: complete session using request body (sessionId) ==========
    @PostMapping("/upload/sessions/complete")
    @Operation(summary = "Complete upload session (using sessionId in request body)")
    public ResponseEntity<?> completeUploadSessionLegacy(@RequestBody CompleteUploadSessionRequest request,
                                                         HttpServletRequest httpRequest) {
        if (request == null || request.getSessionId() == null) {
            return ResponseEntity.badRequest().body(Map.of("success", false, "message", "sessionId is required in request body"));
        }
        StorageUploadResponse response = fileStorageService.completeUploadSession(request.getSessionId(), httpRequest);
        return ResponseEntity.ok(response);
    }

    // ========== Download ==========
    @GetMapping("/download/{fileId}")
    @Operation(summary = "Download file by ID")
    public ResponseEntity<org.springframework.core.io.Resource> download(@PathVariable UUID fileId) {
        StorageDownloadResponse download = fileStorageService.downloadFile(fileId);
        InputStreamResource resource = new InputStreamResource(download.getInputStream());
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + download.getFileName() + "\"")
                .contentType(MediaType.parseMediaType(download.getContentType()))
                .contentLength(download.getFileSize())
                .body(resource);
    }

    @GetMapping("/files/{bucketName}/{*objectKey}")
    @Operation(summary = "Download file by bucket and object key")
    public ResponseEntity<org.springframework.core.io.Resource> downloadByKey(@PathVariable String bucketName,
                                                                              @PathVariable String objectKey,
                                                                              @RequestParam(required = false) Integer width,
                                                                              @RequestParam(required = false) Integer height,
                                                                              HttpServletRequest request) {
        StorageDownloadResponse download = fileStorageService.downloadByKey(bucketName, objectKey);
        InputStreamResource resource = new InputStreamResource(download.getInputStream());
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + download.getFileName() + "\"")
                .contentType(MediaType.parseMediaType(download.getContentType()))
                .contentLength(download.getFileSize())
                .body(resource);
    }

    @DeleteMapping("/{fileId}")
    @Operation(summary = "Delete file")
    public ResponseEntity<Void> delete(@PathVariable UUID fileId) {
        fileStorageService.deleteFile(fileId);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/url/{fileId}")
    @Operation(summary = "Get file URL")
    public ResponseEntity<String> getFileUrl(@PathVariable UUID fileId) {
        String url = fileStorageService.getFileUrl(fileId);
        return ResponseEntity.ok(url);
    }
}